### PR TITLE
linux: drag and drop v2

### PR DESCRIPTION
**Pull request by SmailBarkouch**
This PR adds support for dragging files into Lockbook, and exporting them to your filesystem.

Dragging a file into Lockbook:
<details>

https://user-images.githubusercontent.com/20663038/129415586-73746d71-5e58-41cf-907f-fc789551d53a.mp4
</details>

Exporting a file out of Lockbook:
<details>

https://user-images.githubusercontent.com/20663038/129415648-dc560ff3-9fcd-439a-bffd-d8964079295a.mp4
</details>

### CODE REVIEW

File: clients/linux/src/filetree.rs [hunk 1]
```diff
@@ -183,38 +216,66 @@ impl FileTree {
...
+                                Err(err) => m.send_err_dialog("moving", err)
+                            }
+                        });
+                    }
+                    URI_TARGET_INFO => {
+                        let parent_id = Uuid::parse_str(tree_iter_value!(model, &parent, 2, String).as_str()).unwrap();
+                        let uris: Vec<String> = s.get_uris().iter().map(|uri| uri.to_string()).collect();
+                        m.send(Msg::ShowDialogImportFile(parent_id, uris))
+
+                        // std::thread::spawn(closure!(m, c => move || {
```
Comment (Parth) [inline clients/linux/src/filetree.rs:255]: rm

File: clients/linux/src/app.rs [hunk 2]
```diff
@@ -926,6 +931,153 @@ impl LbApp {
...
+
+        let mut total = 0;
+        let progress = Rc::new(RefCell::new(-1));
+
+        for uri in &uris {
+            if let Some(path) = uri.strip_prefix(FILE_SCHEME) {
+                total += io::get_children_count(Path::new(&path).to_path_buf())?;
+            } else {
+                return Err(LbError::new_user_err(
+                    "Unsupported uri is used.".to_string(),
```
Comment (Parth) [inline clients/linux/src/app.rs:947]:
```suggestion
                    "Unsupported uri!".to_string(),
```

File: clients/linux/src/app.rs [hunk 3]
```diff
@@ -926,6 +931,153 @@ impl LbApp {
...
+
+        spawn!(self.core as c, self.messenger as m => move || {
+            for uri in uris {
+                if let Some(path) = uri.strip_prefix(FILE_SCHEME) {
+                    if let Err(err) = c.import_file(parent, path.to_string(), Some(Box::new(f.clone()))) {
+                        m.send_err_dialog("Import files", err);
+                        break;
+                    };
+                } else {
+                    panic!("impossible");
```
Comment (Parth) [inline clients/linux/src/app.rs:985]: instead of `panic!` can't we just `return`? You can `eprintln!` as well so we notice if this ever happens.

File: clients/linux/src/app.rs [hunk 4]
```diff
@@ -926,6 +931,153 @@ impl LbApp {
         Ok(())
     }
 
+    fn show_dialog_import_file(&self, parent: Uuid, uris: Vec<String>) -> LbResult<()> {
```
Comment (Parth) [inline clients/linux/src/app.rs:934]:
```suggestion
    fn show_import_file_dialog(&self, parent: Uuid, uris: Vec<String>) -> LbResult<()> {
```

File: clients/linux/src/app.rs [hunk 5]
```diff
@@ -926,6 +931,153 @@ impl LbApp {
...
+                }
+            }
+
+            ch.send(None).unwrap();
+        });
+
+        Ok(())
+    }
+
+    fn show_dialog_export_file(&self) -> LbResult<()> {
```
Comment (Parth) [inline clients/linux/src/app.rs:995]:
```suggestion
    fn show_import_file_dialog(&self) -> LbResult<()> {
```

File: clients/linux/src/backend.rs [hunk 6]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
...
+        parent: Uuid,
+        destination: String,
+        f: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+    ) -> LbResult<()> {
+        import_file(&self.config, parent, PathBuf::from(destination), f).map_err(map_core_err!(ImportFileError,
+            FileAlreadyExists => uerr_dialog!("A file already exists in the path with the same name."),
+            DocumentTreatedAsFolder => uerr_dialog!("A document is being treated as folder."),
+            ParentDoesNotExist => uerr_dialog!("The folder does not exist."),
+            NoAccount => uerr_dialog!("No account found."),
+            BadPath => uerr_dialog!("An invalid path was used."),
```
Comment (Parth) [inline clients/linux/src/backend.rs:248]:
```suggestion
            BadPath => uerr_dialog!("Invalid path!"),
```

File: clients/linux/src/backend.rs [hunk 7]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
...
+        &self,
+        parent: Uuid,
+        destination: String,
+        f: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+    ) -> LbResult<()> {
+        import_file(&self.config, parent, PathBuf::from(destination), f).map_err(map_core_err!(ImportFileError,
+            FileAlreadyExists => uerr_dialog!("A file already exists in the path with the same name."),
+            DocumentTreatedAsFolder => uerr_dialog!("A document is being treated as folder."),
+            ParentDoesNotExist => uerr_dialog!("The folder does not exist."),
+            NoAccount => uerr_dialog!("No account found."),
```
Comment (Parth) [inline clients/linux/src/backend.rs:247]:
```suggestion
            NoAccount => uerr_dialog!("No account."),
```

File: clients/linux/src/backend.rs [hunk 8]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
         ))
     }
 
+    pub fn import_file(
+        &self,
+        parent: Uuid,
+        destination: String,
```
Comment (Parth) [inline clients/linux/src/backend.rs:240]: Wouldn't this be the "source"?

File: clients/linux/src/backend.rs [hunk 9]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
...
+        f: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+    ) -> LbResult<()> {
+        import_file(&self.config, parent, PathBuf::from(destination), f).map_err(map_core_err!(ImportFileError,
+            FileAlreadyExists => uerr_dialog!("A file already exists in the path with the same name."),
+            DocumentTreatedAsFolder => uerr_dialog!("A document is being treated as folder."),
+            ParentDoesNotExist => uerr_dialog!("The folder does not exist."),
```
Comment (Parth) [inline clients/linux/src/backend.rs:246]:
```suggestion
            ParentDoesNotExist => uerr_dialog!("Invalid import destination, parent not found."),
```

File: clients/linux/src/backend.rs [hunk 10]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
...
 
+    pub fn import_file(
+        &self,
+        parent: Uuid,
+        destination: String,
+        f: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+    ) -> LbResult<()> {
+        import_file(&self.config, parent, PathBuf::from(destination), f).map_err(map_core_err!(ImportFileError,
+            FileAlreadyExists => uerr_dialog!("A file already exists in the path with the same name."),
+            DocumentTreatedAsFolder => uerr_dialog!("A document is being treated as folder."),
```
Comment (Parth) [inline clients/linux/src/backend.rs:245]:
```suggestion
            DocumentTreatedAsFolder => uerr_dialog!("Invalid import destination, document treated as folder."),
```

File: clients/linux/src/backend.rs [hunk 11]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
...
     }
 
+    pub fn import_file(
+        &self,
+        parent: Uuid,
+        destination: String,
+        f: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+    ) -> LbResult<()> {
+        import_file(&self.config, parent, PathBuf::from(destination), f).map_err(map_core_err!(ImportFileError,
+            FileAlreadyExists => uerr_dialog!("A file already exists in the path with the same name."),
```
Comment (Parth) [inline clients/linux/src/backend.rs:244]:
```suggestion
            FileAlreadyExists => uerr_dialog!("File collision detected during import."),
```

Comment (Parth) [reply clients/linux/src/backend.rs:244]: This seems like *the* error people may get, and we should probably provide a better error for this in the future. Can you make an issue for this?

File: clients/linux/src/backend.rs [hunk 12]
```diff
@@ -233,6 +234,35 @@ impl LbCore {
...
+        &self,
+        parent: Uuid,
+        destination: &str,
+        f: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+    ) -> LbResult<()> {
+        export_file(&self.config, parent, PathBuf::from(destination), f).map_err(map_core_err!(ExportFileError,
+            ParentDoesNotExist => uerr_dialog!("The folder does not exist."),
+            FileAlreadyExistsInDisk => uerr_dialog!("A file already exists in the path with the same name."),
+            NoAccount => uerr_dialog!("No account found."),
+            BadPath => uerr_dialog!("An invalid path was used."),
```
Comment (Parth) [inline clients/linux/src/backend.rs:262]: Just try to use fewer words like I suggest above. But if you're going to use more words, be very specific and contextual regarding it being export related. You know *how* these errors are going to arise. We should be able to debug if someone tells us just the mere string that came up with the error.

**Review by Parth — CHANGES_REQUESTED** [review]

File: core/src/service/import_export_service.rs [hunk 13]
```diff
@@ -0,0 +1,140 @@
+use crate::model::client_conversion;
```
Comment (Parth) [inline core/src/service/import_export_service.rs:1]: This code is good and fine, add some light integration tests in `tests/`.

Maybe one that takes content on disk, imports it, verifies some of the closure properties, exports it out somewhere else, and verify that it's what you expect.

**Review by Parth — CHANGES_REQUESTED** [review]

File: core/src/service/import_export_service.rs (hunk 13 above)
Comment (Parth) [reply core/src/service/import_export_service.rs:1]: Also, isn't there code in CLI that can be deleted now? Can't it use this stuff?

File: clients/linux/src/backend.rs (hunk 11 above)
Comment (SmailBarkouch) [reply clients/linux/src/backend.rs:244]: Sure, [here](https://github.com/lockbook/lockbook/issues/849) it is.

File: clients/linux/src/backend.rs (hunk 8 above)
Comment (SmailBarkouch) [reply clients/linux/src/backend.rs:240]: You are right.

File: core/src/service/import_export_service.rs (hunk 13 above)
Comment (SmailBarkouch) [reply core/src/service/import_export_service.rs:1]: You're right, ill make it use it.

**Comment by Parth** [comment]
@steverusso probably a good time to do a tidy pass here

File: clients/linux/src/app.rs [hunk 14]
```diff
@@ -926,6 +931,154 @@ impl LbApp {
...
+
+            let mut total = 0;
+            let progress = Rc::new(RefCell::new(-1));
+
+            for id in ids.clone() {
+                let meta = self.core.file_by_id(id)?;
+
+                total += match meta.file_type {
+                    FileType::Document => 1,
+                    FileType::Folder => match self.core.get_children_recursively(id) {
```
Comment (steverusso) [inline clients/linux/src/app.rs:1039]:
```suggestion
            for id in &ids {
                let meta = self.core.file_by_id(*id)?;

                total += match meta.file_type {
                    FileType::Document => 1,
                    FileType::Folder => match self.core.get_children_recursively(*id) {
```
Correct me if I'm wrong, but I don't see a good reason to clone here.

File: clients/linux/src/app.rs [hunk 15]
```diff
@@ -926,6 +931,154 @@ impl LbApp {
...
+    fn show_dialog_export_file(&self) -> LbResult<()> {
+        let (selected_tpaths, model) = self.gui.account.sidebar.tree.selected_rows();
+
+        let iters = selected_tpaths
+            .iter()
+            .map(|selected| model.get_iter(selected).unwrap())
+            .collect::<Vec<GtkTreeIter>>();
+        let ids = iters
+            .iter()
+            .map(|iter| Uuid::parse_str(&tree_iter_value!(model, iter, 2, String)).unwrap())
```
Comment (steverusso) [inline clients/linux/src/app.rs:1005]: Why not just map in one go here?

File: clients/linux/src/filetree.rs [hunk 16]
```diff
@@ -162,16 +167,20 @@ impl FileTree {
...
         }
     }
 
     fn on_drag_data_received(
         m: &Messenger,
         c: &Arc<LbCore>,
     ) -> impl Fn(&TreeView, &DragContext, i32, i32, &SelectionData, u32, u32) {
-        closure!(m, c => move |w, d, x, y, _, _, time| {
+        closure!(m, c => move
+            |w, d, x, y, s, info, time| {
```
Comment (steverusso) [inline clients/linux/src/filetree.rs:183]: ?

File: clients/linux/src/filetree.rs [hunk 17]
```diff
@@ -183,38 +192,49 @@ impl FileTree {
...
+                        m.send(Msg::ShowDialogImportFile(parent_id, uris))
+                    }
+                    _ => panic!("unrecognized target info that should not exist")
+                }
 
                 d.drop_finish(true, time);
             }
-        })
+        }
+        )
```
Comment (steverusso) [inline clients/linux/src/filetree.rs:237]: ?

File: clients/linux/src/filetree.rs [hunk 18]
```diff
@@ -183,38 +192,49 @@ impl FileTree {
...
-                        Err(err) => m.send_err_dialog("moving", err)
+                        ids.iter().enumerate().for_each(|(index, id)| {
+                            match c.move_file(id, parent_id) {
+                                Ok(_) => {
+                                    Self::move_iter(&model, &iters[index], &parent, true);
+                                    model.remove(&iters[index]);
+                                }
+                                Err(err) => m.send_err_dialog("moving", err)
+                            }
+                        });
```
Comment (steverusso) [inline clients/linux/src/filetree.rs:224]: I believe this can be simplified (hint: `ids` never even needs to be declared).

File: clients/linux/src/filetree.rs [hunk 19]
```diff
@@ -183,38 +192,49 @@ impl FileTree {
...
                     }
-                });
+                    URI_TARGET_INFO => {
+                        let parent_id = Uuid::parse_str(tree_iter_value!(model, &parent, 2, String).as_str()).unwrap();
```
Comment (steverusso) [inline clients/linux/src/filetree.rs:227]: `parent_id` is the exact same variable in both branches of this match statement, correct? Maybe it can be declared once outside the match.

File: core/tests/import_export_file_tests.rs [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        .unwrap();
+        let root = get_root(config).unwrap();
+
+        let tmp_dir = tempfile::tempdir().unwrap().path().to_path_buf();
+
+        let config_copy = config.clone();
+        let f = move |info: ImportExportFileInfo| {
+            get_file_by_path(&config_copy, info.lockbook_path.as_str()).unwrap();
+            assert!(info.disk_path.exists());
+        };
```
Comment (steverusso) [inline core/tests/import_export_file_tests.rs:35]: I think it's clearer if variables are declared as close as possible to where they are first used (`root` and `f` aren't used until `import_file`.

**Review by steverusso — CHANGES_REQUESTED** [review]

File: clients/linux/src/filetree.rs [hunk 21]
```diff
@@ -183,38 +192,49 @@ impl FileTree {
                         parent = model.get_iter(&path).unwrap();
                     }
                     _ => {
-                        if tree_iter_value!(model, &parent, 3, String) == format!("{:?}", FileType::Document)  {
+                        if tree_iter_value!(model, &parent, 3, String) == format!("{:?}", FileType::Document) {
                             path.up();
                             parent = model.get_iter(&path).unwrap();
                         }
                     }
                 }
```
Comment (steverusso) [inline clients/linux/src/filetree.rs:200]: I know this entirety is not part of this PR right now, but can this match statement not just be an if statement with those conditions (`pos == Before || pos == After) || tree_iter_value!(model, &parent, 3, String) == format!("{:?}", FileType::Document)`)?

**Comment by Parth** [comment]
Also update the PR description with the latest flows, @SmailBarkouch

File: clients/linux/src/filetree.rs [hunk 22]
```diff
@@ -179,52 +179,36 @@ impl FileTree {
...
-                    _ => {
-                        if tree_iter_value!(model, &parent, 3, String) == format!("{:?}", FileType::Document) {
-                            path.up();
-                            parent = model.get_iter(&path).unwrap();
-                        }
-                    }
-                }
-
-                if tree_iter_value!(model, &parent, 3, String) == format!("{:?}", FileType::Document) {
+                if (pos == TreeViewDropPosition::Before || pos == TreeViewDropPosition::After) || tree_iter_value!(model, &parent, 3, String) == format!("{:?}", FileType::Document) {
```
Comment (steverusso) [inline clients/linux/src/filetree.rs]: @SmailBarkouch did you test this and everything worked the same? You got rid of the duplicate code after the match statement, but parent is modified in the match statement after `path.up()`, and then the duplicate code is run again. Now that's not the case, so I just wanted to make sure that nothing broke.

File: clients/linux/src/filetree.rs [hunk 23]
```diff
@@ -179,52 +179,36 @@ impl FileTree {
...
+
                 match info {
                     LOCKBOOK_FILES_TARGET_INFO => {
                         let (paths, _) = w.get_selection().get_selected_rows();
 
-                        let iters = paths.iter().map(|selected| model.get_iter(selected).unwrap()).collect::<Vec<TreeIter>>();
-                        let ids = iters.iter().map(|iter| Uuid::parse_str(&tree_iter_value!(model, iter, 2, String)).unwrap()).collect::<Vec<Uuid>>();
-
-                        let parent_id = Uuid::parse_str(tree_iter_value!(model, &parent, 2, String).as_str()).unwrap();
+                        for selected in paths {
```
Comment (steverusso) [inline clients/linux/src/filetree.rs:198]: What I meant was after declaring `iters`, you could do `iters.iter().map().enumerate().foreach().`, but this works too, whichever one you like better.

File: clients/linux/src/filetree.rs (hunk 22 above)
Comment (SmailBarkouch) [reply clients/linux/src/filetree.rs]: Yeah I get what you mean. The duplicate code was harmless since the `path.up()` statements in the match always made parent into a folder. So when it got to the if statement (the duplicate code) outside of the match it never got to run since the parent was already a folder. I tested it again just to be sure and there are no problems.

**Review by steverusso — CHANGES_REQUESTED** [review]

File: clients/linux/src/app.rs [hunk 24]
```diff
@@ -926,6 +930,156 @@ impl LbApp {
...
+        let (d, disk_lbl, lb_lbl, prog_lbl, pbar) = self.gui.new_import_export_dialog(true);
+
+        const FILE_SCHEME: &str = "file://";
+
+        let mut total = 0;
+        let progress = Rc::new(RefCell::new(-1));
+
+        for uri in &uris {
+            if let Some(path) = uri.strip_prefix(FILE_SCHEME) {
+                total += util::io::get_children_count(Path::new(&path).to_path_buf())?;
```
[the remark this answers is not in the document]
Comment (SmailBarkouch) [reply clients/linux/src/app.rs:943]: You're right.

File: clients/linux/src/app.rs [hunk 25]
```diff
@@ -926,6 +930,156 @@ impl LbApp {
...
+                Some(info) => {
+                    lb_lbl.set_text(&info.lockbook_path);
+                    disk_lbl.set_text(&format!("{}", info.disk_path.display()));
+                    prog_lbl.set_text(&format!("{}/{}", *progress.borrow(), total));
+                }
+            }
+            glib::Continue(true)
+        });
+
+        let f = closure!(ch => move |progress: ImportExportFileInfo| {
```
[the remark this answers is not in the document]
Comment (SmailBarkouch) [reply clients/linux/src/app.rs:972]: Yeah I'm not too sure what I could use to describe it, any suggestions?

[the remark this answers is not in the document]
Comment (Parth) [reply clients/linux/src/app.rs:972]: how about progress_closure?

[the remark this answers is not in the document]
Comment (steverusso) [reply clients/linux/src/app.rs:972]: Whatever the point of it is minus "closure" because that's obvious. Why is it passed all the way to core? Is it to `update_progress`?

[the remark this answers is not in the document]
Comment (Parth) [reply clients/linux/src/app.rs:976]: yeah or `import_progress`

File: core/src/service/import_export_service.rs [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+    config: &Config,
+    parent: Uuid,
+    source: PathBuf,
+    edit: bool,
+    import_progress: Option<Box<dyn Fn(ImportExportFileInfo)>>,
+) -> Result<(), CoreError> {
+    import_file_recursively(
+        config,
+        &source,
+        path_service::get_path_by_id(config, parent)?.as_str(),
```
Comment (steverusso) [inline core/src/service/import_export_service.rs:29]:
```suggestion
        &path_service::get_path_by_id(config, parent)?,
```

File: core/src/service/import_export_service.rs [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+        func(ImportExportFileInfo {
+            disk_path: disk_path.to_path_buf(),
+            lockbook_path: lockbook_path.to_string(),
+        })
+    }
+
+    if disk_path.is_file() {
+        let content = fs::read(&disk_path).map_err(CoreError::from)?;
+        let file_metadata =
+            match path_service::create_at_path(config, lockbook_path_with_new.as_str()) {
```
Comment (steverusso) [inline core/src/service/import_export_service.rs:63]:
```suggestion
            match path_service::create_at_path(config, &lockbook_path_with_new) {
```

File: core/src/service/import_export_service.rs [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+                Ok(file_metadata) => file_metadata,
+                Err(err) => {
+                    if edit && CoreError::PathTaken == err {
+                        path_service::get_by_path(config, lockbook_path_with_new.as_str())?
```
Comment (steverusso) [inline core/src/service/import_export_service.rs:67]:
```suggestion
                        path_service::get_by_path(config, &lockbook_path_with_new)?
```

**Review by steverusso — CHANGES_REQUESTED** [review]

**Review by Parth — APPROVED** [review]

File: clients/linux/src/backend.rs [hunk 29]
```diff
@@ -233,6 +234,42 @@ impl LbCore {
         ))
     }
 
+    pub fn import_file(
+        &self,
+        parent: Uuid,
+        source: String,
```
Comment (steverusso) [inline clients/linux/src/backend.rs:240]: As far as I can tell, this can and should be a `&str`, yes?

File: core/src/lib.rs [hunk 30]
```diff
@@ -760,6 +761,60 @@ pub fn export_drawing_to_disk(
...
+        }
+    })
+}
+
+#[derive(Debug, Serialize, EnumIter)]
+pub enum ExportFileError {
+    NoAccount,
+    ParentDoesNotExist,
+    FileAlreadyExistsInDisk,
+    BadPath,
```
Comment (steverusso) [inline core/src/lib.rs:797]: I don't want to be too nit picky, but the CoreError names for these (`DiskPathTaken` and `DiskPathInvalid`) genuinely seem much clearer to me. Disregard this comment if you disagree, no big deal.

Comment (Parth) [reply core/src/lib.rs:797]: agreed

Comment (SmailBarkouch) [reply core/src/lib.rs]: I agree.

File: core/tests/import_export_file_tests.rs [hunk 31]
```diff
@@ -0,0 +1,134 @@
...
+        import_file, write_document,
+    };
+    use lockbook_models::file_metadata::FileType;
+    use rand::Rng;
+    use uuid::Uuid;
+
+    #[test]
+    fn import_file_successfully() {
+        // new account
+        let config = &temp_config();
```
Comment (steverusso) [inline core/tests/import_export_file_tests.rs:19]: Just take a reference when the variable is being passed somewhere (same in export test below).

**Review by steverusso — CHANGES_REQUESTED** [review]

File: core/tests/import_export_file_tests.rs (hunk 31 above)
Comment (SmailBarkouch) [reply core/tests/import_export_file_tests.rs:19]: I did it this way cause a lot of other tests did as well, just to be consistent.

Comment (steverusso) [reply core/tests/import_export_file_tests.rs:19]: Yea I think it's a good time to start doing it the right way. We'll go back later and clean up the rest, but just for this PR it we can start the change.

Comment (Parth) [reply core/tests/import_export_file_tests.rs:19]: down to not do this in all tests, @SmailBarkouch can you make an issue?

Comment (SmailBarkouch) [reply core/tests/import_export_file_tests.rs:19]: Will update this file with fixes, and added [issue](https://github.com/lockbook/lockbook/issues/853).

**Review by steverusso — APPROVED** [review]
